### PR TITLE
New Path Segment Types

### PR DESCRIPTION
**Pull request by ajohnstonTE**
TL;DR: Adds support for custom path segment types in `MethodUriHandler`.

TL: This should allow any class with a static method named either `fromString` or `valueOf` that accepts a single `String` argument to be used as a method param for a path segment in a `PathUriHandler` method. For example, where previously you might have had:

```java
@Get
@Path("{id}")
public boolean getById(String idStr) {
  UUID id = UUID.fromString(idStr);
  // ...
}
```
you can now do
```java
@Get
@Path("{id}")
public boolean getById(UUID id) {
  // ...
}
```
The `fromString` method takes priority over `valueOf`. As a result, custom enum conversion can be achieved by simply defining a `fromString` method (so that it doesn't use the implicitly defined `valueOf` method). This also has the added benefit of supporting the object-versions of all the primitive classes, such as `Integer` with `Integer.valueOf(String)`.

This is a tiny part of the specification of jax-rs. Relevant jax-rs doc: https://docs.oracle.com/javaee/7/api/javax/ws/rs/PathParam.html
In that page:
> Have a static method named valueOf or fromString that accepts a single String argument (see, for example, Integer.valueOf(String)).

### CODE REVIEW

**Comment by msmith-techempower** [comment]
Okay, Travic-CI, for some reason, lost `gemini` and won't find this PR to test. I'm going to merge in which should kick off a master build - if that fails I will revert and tag this PR again.
